frontend/src/components/pages/TranslationPage.tsx

```diff
@@ -100,6 +124,63 @@ const TranslationPage = () => {
     );
   };
 
+  const onUserInput = async (newContent: string, lineIndex: number) => {
+    const oldContent =
+      translatedStoryLines[arrayIndex(lineIndex)].translatedContent;
+    if (oldContent !== undefined) {
+      updateVersionHistory(oldContent, lineIndex);
+    }
+    onChangeTranslationContent(newContent, lineIndex);
+  };
+
+  const undoChange = () => {
+    if (versionHistoryStack.Undo.length > 0) {
+      const { lineIndex, content: newContent } = versionHistoryStack.Undo[
+        versionHistoryStack.Undo.length - 1
+      ];
+      const oldContent =
+        translatedStoryLines[arrayIndex(lineIndex)].translatedContent;
+      if (oldContent !== newContent) {
+        const newHistory = {
+          Undo: deepCopy(versionHistoryStack.Undo.slice(0, -1)),
+          Redo: [
+            ...deepCopy(versionHistoryStack.Redo),
+            {
+              lineIndex,
+              content: oldContent,
+            },
+          ],
+        };
+        updateVersionHistoryStack(newHistory);
+        onChangeTranslationContent(newContent, lineIndex);
+      }
+    }
+  };
+
+  const redoChange = () => {
+    if (versionHistoryStack.Redo.length > 0) {
+      const { lineIndex, content: newContent } = versionHistoryStack.Redo[
+        versionHistoryStack.Redo.length - 1
+      ];
+      const oldContent =
+        translatedStoryLines[arrayIndex(lineIndex)].translatedContent;
+      if (oldContent !== newContent) {
```

Comment:
In what case would we get `oldContent === newContent`?

Reply:
I think a while back when writing this line I found that the state would double-save when the user first opened the page.